core/src/test/java/org/elasticsearch/bwcompat/MixedRepositoryInteroperabilityIT.java

```diff
@@ -0,0 +1,178 @@
+            final String repoName = "test-repo-" + version;
+            logger.info("-->  creating repository [{}] for version [{}]", repoName, version);
+            createRepository(version, repoName);
+
+            logger.info("--> get the snapshots");
+            final String originalIndex = "index-" + version;
+            final Set<String> indices = Sets.newHashSet(originalIndex);
+            final Set<SnapshotInfo> snapshotInfos = Sets.newHashSet(getSnapshots(repoName));
+            assertThat(snapshotInfos.size(), equalTo(1));
+            SnapshotInfo originalSnapshot = snapshotInfos.iterator().next();
+            assertThat(originalSnapshot.snapshotId(), equalTo(new SnapshotId("test_1", SnapshotId.UNASSIGNED_UUID)));
+            assertThat(Sets.newHashSet(originalSnapshot.indices()), equalTo(indices));
+
+            logger.info("--> restore the original snapshot");
+            final Set<String> restoredIndices = Sets.newHashSet(
+                restoreSnapshot(repoName, originalSnapshot.snapshotId().getName())
+            );
+            assertThat(restoredIndices, equalTo(indices));
+
+            final String snapshotName2 = "test_2";
+            logger.info("--> take a new snapshot of the old index");
+            for (int i = 0; i < 10; i++) {
+                index(originalIndex, "doc", Integer.toString(i), "foo", "new-bar-" + i);
+            }
+            refresh();
+            snapshotInfos.add(createSnapshot(repoName, snapshotName2));
+
+            logger.info("--> get the snapshots with the newly created snapshot [{}]", snapshotName2);
+            Set<SnapshotInfo> snapshotInfosFromRepo = Sets.newHashSet(getSnapshots(repoName));
+            assertThat(snapshotInfosFromRepo, equalTo(snapshotInfos));
+            snapshotInfosFromRepo.forEach(snapshotInfo -> {
+                assertThat(Sets.newHashSet(snapshotInfo.indices()), equalTo(indices));
+            });
+
+            final String snapshotName3 = "test_3";
+            final String indexName2 = "index2";
+            logger.info("--> take a new snapshot with a new index");
+            createIndex(indexName2);
+            indices.add(indexName2);
+            for (int i = 0; i < 10; i++) {
+                index(indexName2, "doc", Integer.toString(i), "foo", "new-bar-" + i);
+            }
+            refresh();
+            snapshotInfos.add(createSnapshot(repoName, snapshotName3));
+
+            logger.info("--> get the snapshots with the newly created snapshot [{}]", snapshotName3);
+            snapshotInfosFromRepo = Sets.newHashSet(getSnapshots(repoName));
+            assertThat(snapshotInfosFromRepo, equalTo(snapshotInfos));
+            snapshotInfosFromRepo.forEach(snapshotInfo -> {
+                if (snapshotInfo.snapshotId().getName().equals(snapshotName3)) {
+                    // only the last snapshot has all the indices
+                    assertThat(Sets.newHashSet(snapshotInfo.indices()), equalTo(indices));
```

Comment:
I see what you mean, testing restore after the repository has had new
snapshots added to it.  I'll add a test for this.

On Tue, May 31, 2016 at 4:50 PM, Igor Motov <notifications@github.com>
wrote:

> In
> core/src/test/java/org/elasticsearch/bwcompat/MixedRepositoryInteroperabilityIT.java
> <https://github.com/elastic/elasticsearch/pull/18228#discussion_r65260020>
> :
>
> > +            logger.info("--> take a new snapshot with a new index");
> > +            createIndex(indexName2);
> > +            indices.add(indexName2);
> > +            for (int i = 0; i < 10; i++) {
> > +                index(indexName2, "doc", Integer.toString(i), "foo", "new-bar-" + i);
> > +            }
> > +            refresh();
> > +            snapshotInfos.add(createSnapshot(repoName, snapshotName3));
> > +
> > +            logger.info("--> get the snapshots with the newly created snapshot [{}]", snapshotName3);
> > +            snapshotInfosFromRepo = Sets.newHashSet(getSnapshots(repoName));
> > +            assertThat(snapshotInfosFromRepo, equalTo(snapshotInfos));
> > +            snapshotInfosFromRepo.forEach(snapshotInfo -> {
> > +                if (snapshotInfo.snapshotId().getName().equals(snapshotName3)) {
> > +                    // only the last snapshot has all the indices
> > +                    assertThat(Sets.newHashSet(snapshotInfo.indices()), equalTo(indices));
>
> Yes. But we don't test that we still can restore after we (potentially)
> messed things by writing a new snapshot. Maybe we didn't actually write
> anything, but instead wiped old snapshot up. We wouldn't know.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18228/files/d973e652a47f5a481de49d99744661f98568eea5#r65260020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjkQd5CtgcygmBcAvsHwF8kz2PncuUwks5qHJ8ygaJpZM4IauFn>
> .
>